File: csharp/StrangeCharactersRefactoring/Characters/CharacterDataParser.cs
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.RegularExpressions;

namespace Characters;

public class CharacterDataParser
{
    private static List<Character> allCharacters = new();
    
    public static void Probably_InitializeFromFile_AndStuff(string? fileName)
    {
        fileName ??= "strange_characters.json";
        string jsonString = File.ReadAllText(fileName, Encoding.UTF8);
        var data = JsonSerializer.Deserialize<List<CharacterData>>(jsonString)!;

        allCharacters = Applesauce(data);
    }

    private static List<Character> Applesauce(List<CharacterData> data)
    {
        List<Character> result = new();
        foreach (var characterData in data)
        {
            var character = new Character(characterData.FirstName, characterData.LastName, characterData.IsMonster);
            result.Add(character);
        }

        foreach (var characterData in data)
        {
            // find nemesis
            if (characterData.Nemesis != null)
            {
                var nemesis = FindCharacter(result, characterData.Nemesis);
                var character = FindCharacter(result, characterData.FirstName);
                character?.SetNemesis(nemesis);
            }

            // add family
            if (characterData.Children != null)
            {
                var character = FindCharacter(result, characterData.FirstName);
                foreach (var childName in characterData.Children)
                {
                    var child = FindCharacter(result, childName);
                    if (child != null)
                        character?.AddChild(child);
                }
            }
        }

        return result;
    }

    public static Character EvaluatePath(string path)
    {
        Character character = null;

        bool hasFamilyName = false;
        string characterName = String.Empty;
        var familyName = String.Empty;
        string tempPathWithoutCurlyBraces = String.Empty;
        string curlyBraces = String.Empty;
        var structureList = path?.Split(new string[] { "/" }, StringSplitOptions.RemoveEmptyEntries);
        for (var i = structureList.Length - 1; i >= 0; i--)
        {
            var localName = String.Empty;
            var localNameWithoutCurlyBraces = String.Empty;
            var familyLocalNameList = structureList[i].Split(":");
            if (familyLocalNameList.Length == 2)
            {
                if (!hasFamilyName)
                {
                    familyName = familyLocalNameList[0];
                    hasFamilyName = true;
                }

                familyName = familyLocalNameList[0];
                localName = familyLocalNameList[1];
            }
            else if (familyLocalNameList.Length == 1)
            {
                localName = familyLocalNameList[0];
            }

            if (i == structureList.Length - 1)
            {
                characterName = localName;
            }

            localNameWithoutCurlyBraces = Regex.Replace(localName, "{[^{]*?}", String.Empty);
            var matches = Regex.Matches(localName, "(.*){([^{]*)}");
            foreach (Match match in matches)
            {
                var groups = match.Groups;
                curlyBraces = groups[2].Value;
                characterName = groups[1].Value;
            }

            tempPathWithoutCurlyBraces = tempPathWithoutCurlyBraces.Insert(0, "/" + localNameWithoutCurlyBraces);
        }

        if (!hasFamilyName)
        {
            character = FindCharacter(allCharacters, characterName);
            if (curlyBraces != null && curlyBraces == "Nemesis")
            {
                return character.Nemesis;
            }
            return character;
        }

        var filteredCharacters = FilterCharactersByFamilyName(familyName, characterName);
        if (filteredCharacters.Any())
        {
            character = FindCharacterWithFamily(filteredCharacters, tempPathWithoutCurlyBraces);
            if (curlyBraces != null && curlyBraces == "Nemesis")
            {
                return character.Nemesis;
            }
        }
        
        return character;
    }

    private static Character? FindCharacterWithFamily(List<Character> filteredCharacters, string tempPathWithoutCurlyBraces)
    {
        var firstName = tempPathWithoutCurlyBraces.Split("/", StringSplitOptions.RemoveEmptyEntries).Last();
        var found = filteredCharacters.FindAll(c => c.FirstName == firstName);
        if (found.Count > 0)
        {
            return found.First();
        }

        return null;
    }

    private static List<Character> FilterCharactersByFamilyName(string familyName, string characterName)
    {
        var found = allCharacters.FindAll(c => c.LastName == familyName);
        if (allCharacters.Exists(c => c.FirstName == characterName))
        {
            return found;
        }

        return new List<Character>();
    }
    

    private static Character? FindCharacter(List<Character> characters, string firstName)
    {
        var found = characters.FindAll(c => c.FirstName == firstName);
        if (found.Count > 0)
        {
            return found.First();
        }
        return null;
    }
    
}

public class CharacterData
{
    [JsonPropertyName("FirstName")] public string FirstName { get; set; }

    [JsonPropertyName("LastName")] public string LastName { get; set; }
    [JsonPropertyName("IsMonster")] public bool IsMonster { get; set; }
    [JsonPropertyName("Nemesis")] public string? Nemesis { get; set; }
    [JsonPropertyName("Children")] public List<string>? Children { get; set; }
}

File: php/src/StrangeCharacters/CharacterDataParser.php
<?php

declare(strict_types=1);

namespace StrangeCharacters;

class CharacterDataParser
{
    private static array $allCharacters = [];
    private static CharacterFinder $characterFinder;

    public static function probably_InitializeFromFile_AndStuff(?string $filename): void
    {
        if ($filename == null) {
            $filename = ROOT_DIR . DIRECTORY_SEPARATOR . "resources" . DIRECTORY_SEPARATOR . "strange_characters.json";
        }

        self::$allCharacters = self::applesauce(json_decode(file_get_contents($filename), false));
        self::$characterFinder = new CharacterFinder(self::$allCharacters);
    }

    private static function applesauce(array $data): array
    {
        $result = [];
        foreach ($data as $characterData) {
            $result[] = Character::withFirstAndLastNameAndMonsterStatus($characterData->FirstName, $characterData->LastName, $characterData->IsMonster);
        }

        foreach ($data as $characterData) {
            // find nemesis
            if (!empty($characterData->Nemesis)) {
                $nemesis = self::findCharacter($characterData->Nemesis, $result);
                $character = self::findCharacter($characterData->FirstName, $result);
                $character->setNemesis($nemesis);
            }

            // add family
            if (!empty($characterData->Children)) {
                $character = self::findCharacter($characterData->FirstName, $result);
                foreach ($characterData->Children as $childName) {
                    $child = self::findCharacter($childName, $result);
                    if ($child != null)
                        $character->addChild($child);
                }
            }
        }

        return $result;
    }

    private static function findCharacter(string $name, array $result): ?Character
    {
        return current(array_filter($result, function (Character $character) use ($name) {
            return $character->firstName === $name;
        }));
    }

    public static function evaluatePath(string $path): ?Character
    {
        $character = null;

        $hasFamilyName = false;
        $characterName = "";
        $familyName = "";
        $tempPathWithoutCurlyBraces = "";
        $curlyBraces = "";
        $structureList = explode("/", $path);
        for ($i = count($structureList) - 1; $i >= 0; $i--) {
            if (empty($structureList[$i]))
                continue;
            $localName = "";
            $familyLocalNameList = explode(":", $structureList[$i]);
            if (count($familyLocalNameList) == 2) {
                if (!$hasFamilyName) {
                    $hasFamilyName = true;
                }

                $familyName = $familyLocalNameList[0];
                $localName = $familyLocalNameList[1];
            } else if (count($familyLocalNameList) == 1) {
                $localName = $familyLocalNameList[0];
            }

            if ($i == count($structureList) - 1) {
                $characterName = $localName;
            }

            $localNameWithoutCurlyBraces = preg_replace("|\{[^{]*?}|", "", $localName);
            $pattern = "|(.*)\{([^{]*)}|";
            $matches = [];
            preg_match($pattern, $localName, $matches);
            if (count($matches) > 0) {
                $curlyBraces = $matches[2];
                $characterName = $matches[1];
            }

            $tempPathWithoutCurlyBraces = "/" . $localNameWithoutCurlyBraces . $tempPathWithoutCurlyBraces;
        }

        if (!$hasFamilyName) {
            $character = self::$characterFinder->findByFirstName($characterName);
            if ($curlyBraces == "Nemesis") {
                return $character->getNemesis();
            }

            return $character;
        }

        $filteredCharacters = self::$characterFinder->findFamilyByLastName($familyName);
        if (!empty($filteredCharacters)) {
            $names = array_filter(explode("/", $tempPathWithoutCurlyBraces), fn(string $name) => !empty($name));
            if (count($names) == 2) {
                $firstName = next($names);
                $candidates = array_filter($filteredCharacters, fn(Character $c) => ($c->firstName == $firstName));
                $character = !empty($candidates) ? current($candidates) : null;
            }
            if ($character != null && $curlyBraces == "Nemesis") {

                return $character->getNemesis();
            }
        }

        return $character;
    }
}